src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1190,8 +1190,16 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             }
         }
 
+        // In some (most?) cases cause.body_id points to actual body, but in some cases
+        // it's a actual definition. According to the comments (e.g. in
+        // librustc_typeck/check/compare_method.rs:compare_predicate_entailment) the latter
+        // is relied upon by some other code. This might (or might not) need cleanup.
+        let body_owner_def_id = match self.tcx.hir().opt_local_def_id(cause.body_id) {
+            Some(def_id) => def_id,
+            None => self.tcx.hir().body_owner_def_id(hir::BodyId{hir_id: cause.body_id}),
+        };
```

Comment:
I'm not sure how sound this is. Hope it's OK at least as a start. Trying to fix all of the code that depends on particular type of `cause.body_id` sounds big enough to be a separate task.

Reply:
This should be fine.

---

I think you can write this as:

```
        let body_owner_def_id = self.tcx.hir().opt_local_def_id(cause.body_id)
            .unwrap_or_else(|| {
                self.tcx.hir().body_owner_def_id(hir::BodyId { hir_id: cause.body_id })
            });
```

Note the small formatting changes as well (spaces around braces).